### PR TITLE
Time Span Type

### PR DESCRIPTION
**Pull request by CalebSerafin**
## What type of PR is this.
* Bug
* Change
3. [x] Enhancement

### What have you changed and why?
* This removes the need to re-design a duration formatter for every mission info, logging systems, and notifications.
* Provides numerous formatting options out the box to fit whatever is desired in one line (No multi-line code on receiving end.).
* It will be used immediatly after merge for telling the admin or commander the last time since gargabe clean on game options statistics menu.
* Added timeSpan class. [negative,days,hours,minutes,seconds,milliseconds,microseconds,nanoseconds]
* Convert seconds into timeSpan.
* Format timeSpan as human readble text.
* Extracted pad 2Digit and 3Digit from systemTime_format_S.

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No
* Yes (Please provide further detail below.)

### secondsToTimeSpan
plug in different scalers and examine output. EG:
```sqf
    [3463463] call A3A_fnc_secondsToTimeSpan;  // [false,40,2,4,23,0,0,0]
    [[3463463] call A3A_fnc_secondsToTimeSpan] call A3A_fnc_timeSpan_format;  // ~"40 Days 2 Hours 4 Minutes 23 Seconds"

    ([[serverTime] call A3A_fnc_secondsToTimeSpan] call A3A_fnc_timeSpan_format) + " since mission started.";  // ~ "9 Hours 30 Minutes 9 Seconds 812 Milliseconds 499 Microseconds 961 Nanoseconds since mission started."
```
### timeSpan_format
Either:
1. Run each of the examlples and check that the output matches.
2. Try break it with **reasonable** inputs.

Examples:
```sqf

DEV_timeSpan = [true,0,0,21,0,0,69,420];
[DEV_timeSpan]                      call A3A_fnc_timeSpan_format;  // "(-) 21 Minutes 69 Microseconds 420 Nanoseconds"
[DEV_timeSpan,1]                    call A3A_fnc_timeSpan_format;  // "(-) 21m 69µs 420ns"
[DEV_timeSpan,2,2]                  call A3A_fnc_timeSpan_format;  // "-0:0:21:0–0:69:420"

// Zeros. Note the negative marker in DEV_timeSpan.
DEV_timeSpan = [true];
[DEV_timeSpan]                      call A3A_fnc_timeSpan_format;  // "(Now)"
[DEV_timeSpan,0,0,true]             call A3A_fnc_timeSpan_format;  // "(+) (Now)"
[DEV_timeSpan,1]                    call A3A_fnc_timeSpan_format;  // "0"
[DEV_timeSpan,1,0,true]             call A3A_fnc_timeSpan_format;  // "(+) 0"
[DEV_timeSpan,0,2,false]            call A3A_fnc_timeSpan_format;  // "0 Days 0 Hours 0 Minutes 0 Seconds 0 Milliseconds 0 Microseconds 0 Nanoseconds"
[DEV_timeSpan,0,2,false,nil,true]   call A3A_fnc_timeSpan_format;  // "00 Days 00 Hours 00 Minutes 00 Seconds 000 Milliseconds 000 Microseconds 000 Nanoseconds"
[DEV_timeSpan,1,2,false]            call A3A_fnc_timeSpan_format;  // "0d 0h 0m 0s 0ms 0µs 0ns"
[DEV_timeSpan,2,2,false]            call A3A_fnc_timeSpan_format;  // "0:0:0:0–0:0:0"
[DEV_timeSpan,2,2,false,nil,true]   call A3A_fnc_timeSpan_format;  // "00:00:00:00–000:000:000"

// Field visibility.
DEV_timeSpan = [false,0,3,54,0,152,0];
[DEV_timeSpan,0]                    call A3A_fnc_timeSpan_format;  // "3 Hours 54 Minutes 152 Milliseconds"
[DEV_timeSpan,0,1]                  call A3A_fnc_timeSpan_format;  // "3 Hours 54 Minutes 0 Seconds 152 Milliseconds"
[DEV_timeSpan,0,2]                  call A3A_fnc_timeSpan_format;  // "0 Days 3 Hours 54 Minutes 0 Seconds 152 Milliseconds 0 Microseconds 0 Nanoseconds"
[DEV_timeSpan,0,2,true]             call A3A_fnc_timeSpan_format;  // "(+) 0 Days 3 Hours 54 Minutes 0 Seconds 152 Milliseconds 0 Microseconds 0 Nanoseconds"

// Slicing.
DEV_timeSpan = [false,0,3,54,0,152,0];
[DEV_timeSpan,0,1,false,2]          call A3A_fnc_timeSpan_format;  // "3 Hours 54 Minutes"
[DEV_timeSpan,0,2,false,[1,1e7]]    call A3A_fnc_timeSpan_format;  // "3 Hours 54 Minutes 0 Seconds 152 Milliseconds 0 Microseconds 0 Nanoseconds"
[DEV_timeSpan,0,2,false,[0,4]]      call A3A_fnc_timeSpan_format;  // "0 Days 3 Hours 54 Minutes 0 Seconds"
[DEV_timeSpan,0,2,false,[1,4]]      call A3A_fnc_timeSpan_format;  // "3 Hours 54 Minutes 0 Seconds"

// Slicing to get digital time.
DEV_timeSpan = [false,0,3,54,0,152,0];
[DEV_timeSpan,2,2,false,[1,4]]      call A3A_fnc_timeSpan_format;  // "3:54:0"
[DEV_timeSpan,2,2,false,[1,4],true] call A3A_fnc_timeSpan_format;  // "03:54:00"
```

### CODE REVIEW

**Comment by CalebSerafin** [comment]
Please critique and comment on how the time spans are presented. If they look funny or if there is a format that should be included.

**Comment by Bob-Murphy** [comment]
Please elaborate where and how the formats should/could be used.
What is the intention behind it other than making other formats available in general.

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    <SCALAR> Symbol Set.    0 are full names. 1 are abbreviations. 2 are condensed colons & en-dash.                                                        [DEFAULT=0]
+    <SCALAR> Show Zeros.    0 will only show non-zero fields. 1 will show in-between zeros between non-zeros. 2 will show all zeros.                        [DEFAULT=0]
+    <BOOL> Show Positive.   false will hide the positive sign. true will allow the positive sign all fields                                                 [DEFAULT=false]
+    <SCALAR> Slice Count.   Number of significant fields to display | <SCALAR,SCALAR> Slice.    First & last Index to be displayed. Days #1, hours #2 ect.. [DEFAULT=1e7]
+    <BOOL> Pad.             All fields will be padded, Days will be padded to 2 characters.                                                                 [DEFAULT=false]
+    <BOOLEAN> Localise.     false for Great British English symbols, true for localised symbols.                                                            [DEFAULT=false]
```
Comment (HakonRydland) [inline A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:15]: inconsistency

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    Note: Days are the largest unit used, because months and years are not fixed amounts and rely on time information to be accurate, and Weeks are not big-enough of a change from days.
+    Note: Sub seconds are used instead of fractions due to float's limited precision.
+
+Arguments:
+    <Timespan> Timespan.    isNegative is index 0, days are index 1, hours are index 2, and smaller units follow in order. May be any amount of fields as long as it starts with isNegative and is in order.
+    <SCALAR> Symbol Set.    0 are full names. 1 are abbreviations. 2 are condensed colons & en-dash.                                                        [DEFAULT=0]
+    <SCALAR> Show Zeros.    0 will only show non-zero fields. 1 will show in-between zeros between non-zeros. 2 will show all zeros.                        [DEFAULT=0]
+    <BOOL> Show Positive.   false will hide the positive sign. true will allow the positive sign all fields                                                 [DEFAULT=false]
+    <SCALAR> Slice Count.   Number of significant fields to display | <SCALAR,SCALAR> Slice.    First & last Index to be displayed. Days #1, hours #2 ect.. [DEFAULT=1e7]
```
Comment (HakonRydland) [inline A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:13]: ~~also array variant~~ literally missed it was there

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    private _lastNonZero = 8;
+    {
+        if (_x isNotEqualTo 0) then {
+            _lastNonZero = _forEachIndex;
+        };
+    } forEach _timeSpan;
+    _timeSpan resize (_lastNonZero + 1);
+};
+
+private _sliceIndexBased = _slice isEqualType [];
```
Comment (HakonRydland) [inline A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:108]: never validates array slice is Array<Scalar,Scalar>

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+private _slicesRemaining = if (_sliceIndexBased) then {1e7} else {_slice};
```
Comment (HakonRydland) [inline A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:109]: should this not be end - start?
also i dont see why _slicesRemaining would be needed at all as you have start and finish

**Review by HakonRydland — CHANGES_REQUESTED** [review]
looks mostly good, would like answers regarding a few things, as for style... i guess look to standard formatting, i certainly dont know of any other that could be needed.

**Comment by CalebSerafin** [comment]
> Please elaborate where and how the formats should/could be used.
> What is the intention behind it other than making other formats available in general.

This removes the need to re-design a duration formatter for every mission info, logging systems, and notifications.
Provides numerous formatting options out the box to fit whatever is desired in one line (No multi-line code on receiving end.).
It will be used immediatly after merge for telling the admin or commander the last time since gargabe clean on game options statistics menu.

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf (hunk 1 above)
Comment (CalebSerafin) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf]: Fixed

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf (hunk 3 above)
Comment (CalebSerafin) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:117]: Moved it to make it more obvious.

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf (hunk 4 above)
Comment (CalebSerafin) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf]: I added some source comments. Basically there are two mode, number of significan fields to display, or slice from start index to end index.

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf (hunk 2 above)
Comment (CalebSerafin) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf]: Moved it to make it more obvious.

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf (hunk 3 above)
Comment (CalebSerafin) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:117]: Each now validated with `param`

**Comment by CalebSerafin** [comment]
@Bob-Murphy Placeholder string table entries have been added for each time field.

File: A3-Antistasi/functions/Time/fn_timeSpan_format.sqf (hunk 3 above)
Comment (HakonRydland) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:117]: still only default, missing the validation [0]

Comment (CalebSerafin) [reply A3-Antistasi/functions/Time/fn_timeSpan_format.sqf:117]: Ah lol, should be good now 👍.

**Review by HakonRydland — APPROVED** [review]

**Comment by HakonRydland** [comment]
yea looks good now

**Pull request merged**